### PR TITLE
Use a version of `postgresql-client` that matches the server version

### LINKED ISSUE #2680: pg_dump: error: aborting because of server version mismatch
Postgres 12 is deprecated and is scheduled for removal in GitLab 16.0.
https://docs.gitlab.com/ee/update/deprecations.html#postgresql-12-deprecated

Using this image with Postgres 13.8 (recommend/default for new installations) will end up in none working backups.  
Database dump (rake task) will fail with pg_dump missmatch.

```
pg_dump: error: server version: 13.8; pg_dump version: 12.13 (Ubuntu 12.13-1.pgdg20.04+1)
pg_dump: error: aborting because of server version mismatch
```

Due to the fact that this image use "postgresql-client-12" there is no pg_dump/restore available for 13.
https://github.com/sameersbn/docker-gitlab/blob/master/Dockerfile

Maybe we need to expose a new ENV to manage this and have both versions available or just make a breaking change?

### PR DESCRIPTION
**Pull request by kkimurak**
Based on #2681 
Close #2680
Related #2572

Current installation uses `postgresql-client-12`. If the database is newer than client, the backup reports an error (but not fail and generates a tar archive which contains empty `database.sql.gz`).

This PR contains following changes.

- Remove the version specification of postgresql-client installed at build time.
  This is a virtual package, intended to install the latest major version of the client that exists at the time of the build.
- Add an additional startup process to check the version of postgresql server, install corresponding version of client, and generate configuration file to switch client version.

### CODE REVIEW

**Comment by kkimurak** [comment]
Discussion:

- [ ] During scheduled backup, we put [whole log to a file](https://github.com/sameersbn/docker-gitlab/blob/213172e9d441050080bae4c340d9a797025643da/assets/runtime/functions#L832-L834). Do we need to pipe error messages to stdout so that we can see the error from `docker log`?
- [ ] Do we need to uninstall unused version of `postgresql-client`?

**Comment by kkimurak** [comment]
cc @sachilles , sorry to ping you, but you probably need this too - This PR will free you from creating branches with different `postgresql-client` versions and building by yourself.

**Comment by sachilles** [comment]
@kkimurak Thanks for your contribution. I'll check your proposal. However, I did not use virtual packages so far. Instead I tried to use the correct postgresql version that is (in my case) almost bleeding edge.

**Comment by sachilles** [comment]
So far, it looks good to me. But please give me some moment to think about it.

**Comment by kkimurak** [comment]
I don't mean to rush you. Deadline is a v16 release (probably a regular release in May 2023), so you can review it when you have time. 
I'm really thank you for your usual work as most active maintainer these years.

**Comment by kkimurak** [comment]
I just rebased onto current master (15.8.2).

**Comment by kkimurak** [comment]
rebased onto current master (15.9.3)

**Comment by kkimurak** [comment]
rebased onto 15.10.3 anyway

**Comment by kkimurak** [comment]
@sachilles Sorry for late, I have updated to follow your suggestion. It's ready for review.

- Multiple versions of `postgresql-client` will be installed on build and shipped with image
- Check the server side version when starting the container, and generate .postgresqlrc to use the corresponding version of client (No installation will be executed runtime)
- Unused `postgresql-client-*` will be removed as well
  In this context, "Unused" means "not listed in .postgresqlrc"

All clients that will be supported (currently 12, 13, 14, 15) will be installed at build time to satisfy the desire to not have an internet connection at runtime.
We need to keep updating the Dockerfile in the future, as new postgresql versions are released or something else. 

If there is no client that exactly matches the major version of the server side, the newest version of the installed client will be used. There are two possible scenarios:
- If the user is using postgresql which is not currently supported (for example: version earlier than 11) 
- If the user try to use a version of postgresql server that did not exist at the time of build

Here is an example `docker log` in my test environment with server 14.7-1. I have injected a code to uninstall `postgresql-client-14` before `gitlab_generate_postgresqlrc()`, so we can see warning about version mismatch.

````
Configuring gitlab::database
Configuring /home/git/.postgresqlrc to avoid version mismatch on dumping
- Detected server version: 140007
gitlab_generate_postgresqlrc(): WARNING - Suitable client not installed. postgresql-client-15 will be used but you may face issue (database in backup will be empty, for example)
- Generating /home/git/.postgresqlrc
15 postgresql:5432 gitlabhq_production
- Uninstalling unused client(s): postgresql-client-12 postgresql-client-13 
````

**Comment by kkimurak** [comment]
Is there already demand for postgresql-client-16? If so, would we like to include it in this PR?

**Comment by kkimurak** [comment]
Rebased onto 15.11.3 - this should be the final release before 15.12.0 which ends support for postgresql 12.x (unless gitlab does an urgent security release).  

Action is required before that, whether or not this PR is merged. There should be no reason to do a release that obviously wouldn't work.

**Comment by sachilles** [comment]
@kkimurak 

Sorry for being late.

> Is there already demand for postgresql-client-16? If so, would we like to include it in this PR?

I guess this is currently not the case. As far as I know PostgreSQL 16 will be released in the third quarter of 2023.

I'll test your PR once more. So far it looks good, but anyway we have to update the README notes.

**Comment by kkimurak** [comment]
Thank you for your time, @sachilles .

 > I guess this is currently not the case. As far as I know PostgreSQL 16 will be released in the third quarter of 2023.

OK thank you. Let's keep an eye on future releases and deprecation notes.

> but anyway we have to update the README notes.

I've reviewed it briefly, but what part to be updated as a part of this Pull Request? I see a lot of outdated descriptions in the [section about postgresql](https://github.com/sameersbn/docker-gitlab/blob/d35c73ce724e4a585962e135f685cd9e9c3b00e6/README.md#postgresql).

**Comment by sachilles** [comment]
Hi,

you're right. I guess there has to be some kind of maintenance on the documentation. SInce I've the feeling that the documentation has only grown over time.

I think that at least in the release notes the incorporation of all PostgreSQL client versions in the image (that are supported by GitLab) should be mentioned.

The section about the DB connection in the README file could be done later in detail as well. However, I think a general note in this section about the PostgreSQL client versions in the image are also helpful.

**Comment by kkimurak** [comment]
Thank you for your comment. I just wrote simple description about internal behavior (that switches client version on startup). Any feedbacks are welcome.

**Review by sachilles — CHANGES_REQUESTED** [review]
Please check my comment on the part of the README file. If you are not complaining, I kindly ask you to insert the modified text. After that I will merge the PR and prepare the new releases (since the have been published already).

Anycase, I thank you for this very valuable contribution.

File: README.md
Comment (kkimurak) [reply README.md]: Thank you for your suggestion. All accepted (correcting code block markers).
Additionally, added a note when the major version of any client version does not match the server.

**Review by sachilles — APPROVED** [review]
LGTM

**Comment by kkimurak** [comment]
@f-io You're right, I should have used the purge command, but completely forgot that. Fix PR will be soon..

**Pull request merged**
